### PR TITLE
feat(api_keys): add api keys route to api v2

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] New feature

## Description

This PR adds the api key routes to api v2. In addition to it the authentication layer is refactored for admin apis to always get the merchant account and key store. This is done in order to maintain consistency across all admin apis.

In v2 routes, the `merchant_id` is extracted from the `x-merchant-id` header. This logic has been added in the authentication layer.

## How did you test it?

- Postman tests
- Api v2 postman collection

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/routes/app.rs [hunk 1]
```diff
@@ -1405,7 +1405,27 @@ impl Poll {
...
+impl ApiKeys {
+    pub fn server(state: AppState) -> Scope {
+        web::scope("/v2/api_keys")
+            .app_data(web::Data::new(state))
+            .service(web::resource("").route(web::post().to(api_key_create)))
+            .service(web::resource("/list").route(web::get().to(api_key_list)))
+            .service(
+                web::resource("/{key_id}")
+                    .route(web::get().to(api_key_retrieve))
+                    .route(web::post().to(api_key_update))
```
Comment (SanchithHegde) [inline crates/router/src/routes/app.rs:1418]: Shouldn't this be `PUT` method to maintain consistency with the other v2 update APIs?

```suggestion
                    .route(web::put().to(api_key_update))
```

File: crates/router/src/routes/admin.rs [hunk 2]
```diff
@@ -644,7 +516,7 @@ pub async fn connector_update(
             )
         },
         auth::auth_type(
-            &auth::AdminApiAuthWithMerchantId::default(),
+            &auth::AdminApiAuthWithMerchantIdFromHeader,
```
Comment (SanchithHegde) [inline crates/router/src/routes/admin.rs:519]: Isn't this going to affect existing Postman or Cypress tests? Shouldn't we be updating the v2 merchant connector update route instead?

File: crates/router/src/services/authentication.rs [hunk 3]
```diff
@@ -693,10 +693,10 @@ where
 }
 
 #[derive(Debug, Default)]
```
Comment (SanchithHegde) [inline crates/router/src/services/authentication.rs:695]: Nit: I think we can remove the `Default` derive here.

File: crates/router/src/routes/admin.rs (hunk 2 above)
Comment (Narayanbhat166) [reply crates/router/src/routes/admin.rs:519]: The `default()` checks the merchant id from header. I have just renamed this

**Review by SanchithHegde — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

File: crates/router/src/services/authentication.rs [hunk 4]
```diff
@@ -755,6 +742,113 @@ where
...
+                        message: format!("`{}` header is invalid", headers::X_MERCHANT_ID),
+                    },
+                )
+            })
+    }
+}
+
+/// Get the merchant-id from `x-merchant-id` header
+#[derive(Debug)]
+pub struct AdminApiAuthWithMerchantIdFromHeader;
```
Comment (hrithikesh026) [inline crates/router/src/services/authentication.rs:803]: This is already present.
pub struct AdminApiAuthWithMerchantId(AdminApiAuth);

File: crates/api_models/src/admin.rs [hunk 5]
```diff
@@ -767,10 +767,6 @@ pub struct MerchantConnectorCreate {
     // By default the ConnectorStatus is Active
     pub status: Option<api_enums::ConnectorStatus>,
 
-    /// The identifier for the Merchant Account
-    #[schema(value_type = String, max_length = 64, min_length = 1, example = "y3oqhf46pyzuxjbcn2giaqnb44")]
-    pub merchant_id: id_type::MerchantId,
-
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/api_models/src/admin.rs:773]: This has been removed in v2 feature only

**Pull request merged**